### PR TITLE
Niki/receivers grpc

### PR DESCRIPTION
**Pull request by niki-joanne**
## Changes made in this PR
Please describe what change you've made in this branch.
Added the grpc implementation for CreateReceiver

## Corresponding noted Ruby branch
https://github.com/ParamountDataManagement/paramount-api/tree/niki/golang-receiver-implementation

## TODO - changes that needs to be made
Write grpc implementation for remaining receiver functions 
                                                                                        
## Implementation Checklist
Please check following check list to ensure the tasks are considered and completed in this branch.

#### API Design
- [x] Defined gRPC API in proto file
- [x] regenerated go code using `make proto_generate` command

#### API implementation
- [x] Implemented API in corresponding gRPC server
- [x] Added tests in gRPC layer

#### Domain Service and DB query Implementation
- [x] Added new query in dbstore layer
- [x] Added tests in db store layer
                                                                                       
- [x] Implemented functionality in Domain Service Layer
- [x] Added test in Domain Service Layer

- [x] Created a new branch in paramount-api, with comments there indicating the functionality in Ruby that's re-implemented in Golang

#### Postgres Layer - identify and potentially add db migration to create index for fast querying.
- [ ] Verified and potentially created index for the newly added query.

### CODE REVIEW

File: migrations/20220128193135_CreateReceivers.up.sql [hunk 1]
```diff
@@ -12,7 +12,7 @@ create table if not exists receivers (
     town varchar(250),
     country varchar (250),
     primary_contact_id integer,
-    revenue_range varchar (100),      --TODO: lhan move to ENUM
+    revenue_range varchar (100) not null default 'unknown',      --TODO: lhan move to ENUM
```
Comment (lhan-the-real) [inline migrations/20220128193135_CreateReceivers.up.sql:15]: We don't need this change. Specifically, as we discussed via call, separate layer has separate representation of an entity. API layer using proto message, golang application layer using struct, database layer using table. 
We will write code to convert and map different layer of entity representation, but we don't  mix different layer's entity representation.

revenue_range can be empty, can be null. Let's revert this db migration change.

File: internal/domain/receivers/receiversStore_test.go [hunk 2]
```diff
@@ -130,7 +126,7 @@ var _ = Describe("ReceiversStore", func() {
 
 		Context("update a nonexistent receiver", func() {
 			It("should throw errors", func() {
-				updatereceiverInfo.ID = -1
+				updatereceiverInfo.ID = -100
```
Comment (lhan-the-real) [inline internal/domain/receivers/receiversStore_test.go:129]: This seems to be an unnecessary change. Why we change it to -100?

File: internal/domain/receivers/receiversStore.go [hunk 3]
```diff
@@ -53,40 +53,16 @@ type receiversStore struct {
...
+	createdReceiver = Receiver{
+		Name: receiverName,
 	}
 
-	err = sqlStatement.Get(&createdReceiver, newReceiver)
+	query := fmt.Sprintf(`insert into receivers(name) values($1) returning %s;`, sqlReceiversAttrs)
+	err := a.db.Get(&createdReceiver, query, receiverName)
 	if err != nil {
-		return Receiver{}, pdmerr.NewParent(err, "ReceiversStore: failed to CreateReceiver from DB\n")
+		return Receiver{}, pdmerr.NewParent(err, "ReceiversStore: failed to CreateReceiver from DB - prepared sql query statement fails\n")
```
Comment (lhan-the-real) [inline internal/domain/receivers/receiversStore.go:65]: The error message needs to be updated. We are not using prepared sql query statement anymore.

File: internal/domain/receivers/receiversStore.go [hunk 4]
```diff
@@ -53,40 +53,16 @@ type receiversStore struct {
...
-				:linkedin_name, 
-				:amazon_seller_name, 
-				:ebay_seller_name, 
-				:is_visible)
-				returning %s;`, sqlReceiversAttrs)
-	sqlStatement, err := a.db.PrepareNamed(query)
-	if err != nil {
-		return Receiver{}, pdmerr.NewParent(err, "ReceiversStore: failed to CreateReceiver from DB - prepared sql query statement fails\n")
+	createdReceiver = Receiver{
+		Name: receiverName,
```
Comment (lhan-the-real) [inline internal/domain/receivers/receiversStore.go:59]: What's the purpose for assigning a Receiver struct to createdReceiver?

File: internal/domain/receivers/receiversDomainService.go [hunk 5]
```diff
@@ -11,31 +11,41 @@ import (
...
+	Name              string           `db:"name"`
+	Description       sql.NullString   `db:"description"`
+	Address1          sql.NullString   `db:"address_1"`
+	Address2          sql.NullString   `db:"address_2"`
+	Zip               sql.NullString   `db:"zip"`
+	State             sql.NullString   `db:"state"`
+	Town              sql.NullString   `db:"town"`
+	Country           sql.NullString   `db:"country"`
+	PrimaryContactID  sql.NullInt32    `db:"primary_contact_id"`
+	RevenueRange      RevenueRangeEnum `db:"revenue_range"`
```
Comment (lhan-the-real) [inline internal/domain/receivers/receiversDomainService.go:39]: Let's still use sql.NullString here.

File: internal/domain/receivers/receiversDomainService.go [hunk 6]
```diff
@@ -11,31 +11,41 @@ import (
 
 var loggerNameValue = "internal/domain/receivers"
 
+type RevenueRangeEnum string
+
+const (
+	REVENUE_RANGE_UNKNOWN RevenueRangeEnum = "unknown"
```
Comment (lhan-the-real) [inline internal/domain/receivers/receiversDomainService.go:17]: Change this line to :
```
	REVENUE_RANGE_UNKNOWN RevenueRangeEnum = ""
```

File: internal/domain/receivers/receiversDomainService.go [hunk 7]
```diff
@@ -11,31 +11,41 @@ import (
 
 var loggerNameValue = "internal/domain/receivers"
 
+type RevenueRangeEnum string
```
Comment (lhan-the-real) [inline internal/domain/receivers/receiversDomainService.go:14]: Rename `RevenueRangeEnum` to `RevenueRange`

File: apis/grpcservice/grpcutils/structToProto.go [hunk 8]
```diff
@@ -58,3 +60,84 @@ func getAccountStatusProto(status accounts.AccountStatus) pb.Account_AccountStat
...
+	case receivers.REVENUE_RANGE_1_5:
+		return rpb.Receiver_REVENUE_RANGE_1_5
+	case receivers.REVENUE_RANGE_5_50:
+		return rpb.Receiver_REVENUE_RANGE_5_50
+	case receivers.REVENUE_RANGE_50_PLUS:
+		return rpb.Receiver_REVENUE_RANGE_50_PLUS
+	}
+
+	return rpb.Receiver_REVENUE_RANGE_UNKNOWN
+}
```
Comment (lhan-the-real) [inline apis/grpcservice/grpcutils/structToProto.go:143]: Let's make the usage of the revenueRangeEnum here:
Change the implementation to 
```
func getReceiverRevenueRangeProto(revenueRange sql.NullString) rpb.Receiver_RevenueRange {
	if !revenueRange.Valid {
		return rpb.Receiver_REVENUE_RANGE_UNKNOWN
	}
	switch receivers.RevenueRangeEnum(revenueRange.String) {
	case receivers.REVENUE_RANGE_0_1:
		return rpb.Receiver_REVENUE_RANGE_0_1
	case receivers.REVENUE_RANGE_1_5:
		return rpb.Receiver_REVENUE_RANGE_1_5
	case receivers.REVENUE_RANGE_5_50:
		return rpb.Receiver_REVENUE_RANGE_5_50
	case receivers.REVENUE_RANGE_50_PLUS:
		return rpb.Receiver_REVENUE_RANGE_50_PLUS
	}

	return rpb.Receiver_REVENUE_RANGE_UNKNOWN
}

```

Comment (lhan-the-real) [reply apis/grpcservice/grpcutils/structToProto.go:143]: Note the input is switched to `sql.NullString` which is matching the golang receiver struct. revenueRange type representation. 

In the conversion here, we mainly use the enum to guard against invalid value of the revenueRange to be converted wrongly from golang struct to grpc enum.

File: apis/grpcservice/grpcreceivers/receiversGRPCServer_test.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+			Name: "test_name",
+		}
+	})
+
+	AfterEach(func() {
+		db.Query("truncate table receivers cascade")
+	})
+
+	Describe("CreateReceiver", func() {
+		It("Return Receiver Info", func() {
```
Comment (lhan-the-real) [inline apis/grpcservice/grpcreceivers/receiversGRPCServer_test.go:41]: We should give it a `unhappy path` test case.
When could this CreateReceiver go wrong? And in that scenario, what's the expected response?

According to the above consideration, since there's only receiver name as attribute in the request, we can add a test case with `receiverName == ""` And test accordingly.

File: internal/domain/receivers/receiversStore_test.go (hunk 2 above)
Comment (niki-joanne) [reply internal/domain/receivers/receiversStore_test.go:129]: Accidental change, when I was playing around with some things, and I forgot to change it back.
My bad, will revert back now.

File: internal/domain/receivers/receiversStore.go (hunk 4 above)
Comment (niki-joanne) [reply internal/domain/receivers/receiversStore.go:59]: Something I did in the initial phases, and forgot to remove. My apologies, getting rid of it now.

File: apis/grpcservice/grpcreceivers/receiversGRPCServer_test.go (hunk 9 above)
Comment (niki-joanne) [reply apis/grpcservice/grpcreceivers/receiversGRPCServer_test.go]: I've added that test case, please verify if I've done so correctly
I was also thinking I could add one for a duplicate name, if need be?

Comment (lhan-the-real) [reply apis/grpcservice/grpcreceivers/receiversGRPCServer_test.go]: Good point. Let's add this test.

File: internal/domain/receivers/receiversStore.go [hunk 10]
```diff
@@ -13,7 +13,7 @@ import (
 // ReceiversStore - interface for receiversStore db functions
 type ReceiversStore interface {
 	// TODO: unimplemented real functionalities, need to update the function signature when implementing the real functionality.
-	CreateReceiver(newReceiver Receiver) (createdReceiver Receiver, err error)
+	CreateReceiver(receiverName string) (createdReceiver Receiver, err error)
```
Comment (lhan-the-real) [inline internal/domain/receivers/receiversStore.go:16]: We can change this return type to `createdReceiver *Receiver, err error `

File: internal/domain/receivers/receiversStore.go [hunk 11]
```diff
@@ -53,40 +53,12 @@ type receiversStore struct {
...
+	query := fmt.Sprintf(`insert into receivers(name) values($1) returning %s;`, sqlReceiversAttrs)
+	err := a.db.Get(&createdReceiver, query, receiverName)
 	if err != nil {
-		return Receiver{}, pdmerr.NewParent(err, "ReceiversStore: failed to CreateReceiver from DB - prepared sql query statement fails\n")
-	}
-
-	err = sqlStatement.Get(&createdReceiver, newReceiver)
-	if err != nil {
-		return Receiver{}, pdmerr.NewParent(err, "ReceiversStore: failed to CreateReceiver from DB\n")
+		return Receiver{}, pdmerr.NewParent(err, "ReceiversStore: failed to CreateReceiver from DB - sql query statement fails\n")
```
Comment (lhan-the-real) [inline internal/domain/receivers/receiversStore.go:61]: If we change the return type to `*Receiver, error` then here we can do 
```
return nil, pdmerr.NewParent()
```

**Comment by lhan-the-real** [comment]
Nice! Let's merge latest main branch into this, and then we are ready to merge it!
